src/python/pants/engine/unions_test.py

```diff
@@ -19,3 +23,112 @@ class B:
     assert UnionMembership.from_rules([UnionRule(Base, A), UnionRule(Base, B)]) == UnionMembership(
         {Base: FrozenOrderedSet([A, B])}
     )
+
+
+def test_unions_using_protocols() -> None:
+    # Setup different versions of Union base classes, show casing UnionBase vs Protocol (w/ and w/o
+    # @runtime_checkable).
+
+    class Base(UnionBase):
+        def frobnicate(self) -> bool:
+            ...
+
+    # This will be the line to go with, using Protocol directly. The runtime_checkable decorator
+    # could be made optional, especially if using attributes, which it is not compatible with.. I
+    # was hoping for something cleaner, but maybe not too bad, considering the hackish workarounds
+    # required to make it work otherwise.
+    @runtime_checkable
+    class Base2(Protocol):
```

Comment:
I think that in order to replace unions, this will require one more explicit opt-in: we need to be able to detect that something is a union base for: https://github.com/pantsbuild/pants/blob/731c855630a9eed6c68fdf6129541dc6a69ec3e1/src/python/pants/engine/internals/scheduler.py#L658-L666

...and I'm not sure that extending `Protocol` is sufficient to indicate that you're not using a concrete class in that position? ... maybe it is though, if there is a convention to never actually ever extend `Protocol`? But I don't think it's prohibited...

Reply:
Can't line 659 just be switched (today - independent of this WIP) to using the exact `the_get.input_type` as a key into `union_membership` instead of `union.is_instance`?

Reply:
Oh. Duh. Yes.

Reply:
This location needs patching too..
https://github.com/pantsbuild/pants/blob/731c855630a9eed6c68fdf6129541dc6a69ec3e1/src/python/pants/engine/internals/selectors.py#L202-L211

Reply:
> Can't line 659 just be switched (today - independent of this WIP) to using the exact the_get.input_type as a key into union_membership instead of union.is_instance?

No, it does not work, as it changes the semantics of when `add_get_edge` is called. Consider a union `input_type` but that has not been registered, then we end up calling `add_get_edge(output_type, input_type)` rather than not calling `add_get_edge` at all.

I think a better approach is to move the union test into a dedicated function in `unions.py`. Oh, wait, there already is such a function, we should only use it :P